Fix Invertd

eclipse0922 opened the pull request:
Fixes #8396.

### Description

This PR fixes a regression where `Invertd` and direct `Compose.inverse()` calls could fail when later post-processing steps had appended unrelated entries to the same transform history.

Root cause:
- inverse execution assumed that the most recent history entry had to belong to the transform currently being inverted
- once post-processing appended additional history entries, inverse would pop from the tail, hit the wrong transform, and raise an ID mismatch

Solution:
- centralize inverse matching in the shared `TraceableTransform` history lookup path
- search backward for the most recent matching transform entry instead of assuming the tail entry must match
- pop only the matched preprocessing entry, leaving unrelated post-processing history intact
- remove the group-tracking plumbing that was previously introduced in `Compose`, `TraceKeys`, and `Invertd`

Regression coverage added for:
- the original `Invertd` reproducer from #8396
- multiple unrelated trailing post-processing transforms
- same-class trailing transform history on `MetaTensor`
- trace-key history with a different instance of the same invertible transform class
- direct `Compose.inverse()` with mixed post-processing history

### Types of changes
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

Commit:
Enables automatic transform group tracking for inversion

Addresses an issue where `Invertd` fails when postprocessing
contains invertible transforms before `Invertd` is called.
The solution uses automatic group tracking: `Compose` assigns
its ID to child transforms, allowing `Invertd` to filter and
select only the relevant transforms for inversion.

This ensures correct inversion when multiple transform pipelines
are used or when post-processing steps include invertible transforms.
`TraceableTransform` now stores group information. `Invertd` now
filters transforms by group, falling back to the original
behavior if no group information is present (for backward
compatibility). Adds tests to verify the fix and group isolation.

Commit:
autofix formatting

DCO Remediation Commit for sewon.jeon <sewon.jeon@connecteve.com>

I, sewon.jeon <sewon.jeon@connecteve.com>, hereby add my Signed-off-by to this commit: d97bdd574db39960dd42046ff2b868f9ac091316
I, sewon.jeon <sewon.jeon@connecteve.com>, hereby add my Signed-off-by to this commit: c523e945155f1662eef7e73dc562df7bd812e61d

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

Commit:
fix errors

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

eclipse0922 commented:
Hi @KumoLiu @Nic-Ma @ericspod,  
Sorry for the ping — just checking if you might have time to review this PR.  
It’s been up for a while and all CI checks are green.  
Thanks in advance!

atbenmurray commented:
I can take a look at this.

ericspod commented:
> I can take a look at this.

Please do have a look @atbenmurray, thanks. Sorry @eclipse0922 for the delay but we have a CI issue we need to sort with a patch release anyway.

eclipse0922 committed:
Add Google-style docstrings for transform group helpers

eclipse0922 committed:
Assign group IDs to wrapped transforms in Compose

Extend Compose group tagging to discover traceable transforms wrapped in object attributes and container fields, including common patterns like transform/transforms wrappers.

Preserve nested Compose boundaries by setting group IDs only on the nested Compose instance itself, without traversing its internal pipeline.

Add a regression test covering wrapped transform attributes and nested container traversal.

python -m pip install -U -r requirements-dev.txt

eclipse0922 committed:
DCO Remediation Commit for sewon jeon <irocks0922@gmail.com>

I, sewon jeon <irocks0922@gmail.com>, hereby add my Signed-off-by to this commit: 821bc01667e00bd0ed2658eda5aa9b621b722923
I, sewon jeon <irocks0922@gmail.com>, hereby add my Signed-off-by to this commit: 6e854a8789222d1f1b50112b1ab9a99095aea46f

Signed-off-by: sewon jeon <irocks0922@gmail.com>

ericspod commented on monai/transforms/post/dictionary.py:
```diff
@@ -859,6 +859,29 @@ def __init__(
+        if not filtered:
+            return all_transforms
+
+        return filtered
```
```suggestion
        filtered = [xform for xform in all_transforms if xform.get(TraceKeys.GROUP) == target_group]

        # If no transforms match (backward compatibility), return all transforms
        return filtered if filtered else all_transforms
```

Ignore if you think this is harder to read.

eclipse0922 replied:
Applied in commit 5734cffe5: switched to a list comprehension for `filtered` and kept the backward-compatible fallback `return filtered if filtered else all_transforms`.

ericspod commented on monai/transforms/post/dictionary.py:
```diff
@@ -859,6 +859,29 @@ def __init__(
+        Filter applied_operations to only include transforms from the target Compose instance.
+        Uses automatic group tracking where Compose assigns its ID to child transforms.
+        """
+        from monai.utils import TraceKeys
```
Imports should always be at the top of the source file and sorted according to isort. If you need to import this here to avoid circular importation please note that with a comment.

eclipse0922 replied:
Updated in commit 5734cffe5: moved `TraceKeys` to the module-level imports in `monai/transforms/post/dictionary.py` and removed the local import in `_filter_transforms_by_group`.

ericspod commented on tests/transforms/inverse/test_invertd.py:
```diff
@@ -137,6 +137,224 @@ def test_invert(self):
+            self.assertIn(key, post)
+            print("SUCCESS! Automatic group tracking fixed the bug.")
+            print(f"  Preprocessing group ID: {id(preprocessing)}")
+            print(f"  Postprocessing group ID: {id(postprocessing)}")
```
Tests in general should be silent unless they fail, this keeps the log easier to read so please remove these.

eclipse0922 replied:
Removed the debug `print(...)` statements from `test_invertd_with_postprocessing_transforms` in commit 5734cffe5 so the test remains silent on success.

ericspod commented on tests/transforms/inverse/test_invertd.py:
```diff
@@ -137,6 +137,224 @@ def test_invert(self):
+        The fix uses automatic group tracking where Compose assigns its ID to child transforms.
+        """
+        from monai.data import MetaTensor, create_test_image_2d
+        from monai.transforms.utility.dictionary import Lambdad
```
These imports should all be at the top of the file and not in methods, same for those below.

eclipse0922 replied:
Updated in commit 5734cffe5: hoisted method-local imports to the file top (`MetaTensor`, `create_test_image_2d`, and `Lambdad`) and removed the in-method imports in these new tests.

ericspod commented on tests/transforms/inverse/test_invertd.py:
```diff
@@ -137,6 +137,224 @@ def test_invert(self):
+        # Inverting preprocessing1 should only invert its transforms
+        inverter = Invertd(key, transform=preprocessing1, orig_keys=key)
+        inverted = inverter(pre2)
+        self.assertIsNotNone(inverted)
```
Here it might be good to check the size of items in `inverted` to ensure the correct transforms were inverted only.

eclipse0922 replied:
Added explicit checks in `test_invertd_group_isolation` (commit 5734cffe5): `pre2[key].shape == (1, 40, 40)`, per-group occurrence counts for `preprocessing1`/`preprocessing2` in `applied_operations`, and `inverted[key].shape == (1, 60, 60)`.

ericspod reviewed (changes requested):
Hi @eclipse0922 sorry for the delay on this one. I have had a look through and I think this solution will work. I'm a little worried about how complex our transform implementation is getting with fixes like these being added on, but this looks like a good solution to the problem. I would like @atbenmurray to have a look as well and I made a few comments.

eclipse0922 committed:
address the comments

Signed-off-by: sewon jeon <irocks0922@gmail.com>

eclipse0922 committed:
DCO Remediation Commit for sewon.jeon <irocks0922@gmail.com>

I, sewon.jeon <irocks0922@gmail.com>, hereby add my Signed-off-by to this commit: 5734cffe5dd310c549bdaa71a1ccae6cfc691c91

Signed-off-by: sewon.jeon <irocks0922@gmail.com>

eclipse0922 committed:
Address CodeRabbit follow-up review comments

eclipse0922 commented on monai/transforms/post/dictionary.py:
```diff
@@ -859,6 +859,20 @@ def __init__(
+        filtered = [xform for xform in all_transforms if xform.get(TraceKeys.GROUP) == target_group]
+
+        # If no transforms match (backward compatibility), return all transforms
+        return filtered if filtered else all_transforms
```
Updated in commit f6777abd: `_filter_transforms_by_group` docstring now uses Google-style sections with explicit `Args` and `Returns`, including the `TraceKeys.GROUP` matching/backward-compat behavior.

eclipse0922 commented on monai/transforms/post/dictionary.py:
```diff
@@ -894,8 +908,11 @@ def __call__(self, data: Mapping[Hashable, Any]) -> dict[Hashable, Any]:
+
+                # Automatically filter by Compose instance group ID
+                transform_info = self._filter_transforms_by_group(all_transforms)
             else:
                 transform_info = d[InvertibleTransform.trace_key(orig_key)]
                 meta_info = d.get(orig_meta_key, {})
```
Applied in commit f6777abd: the non-MetaTensor path now also calls `_filter_transforms_by_group(...)` before using `transform_info`, so group filtering is consistent for both metadata-storage paths.

eclipse0922 commented on tests/transforms/inverse/test_invertd.py:
```diff
@@ -137,6 +138,209 @@ def test_invert(self):
+            self.assertIn(key, post)
+        except RuntimeError as e:
+            if "getting the most recently applied invertible transform" in str(e):
+                self.fail(f"Invertd still has the postprocessing transform bug: {e}")
```
Updated in commit f6777abd: removed the `try/except RuntimeError` wrapper in `test_invertd_with_postprocessing_transforms` and now call `post = postprocessing(pre)` directly so any exception fails the test naturally.

eclipse0922 committed:
DCO Remediation Commit for sewon.jeon <irocks0922@gmail.com>

I, sewon.jeon <irocks0922@gmail.com>, hereby add my Signed-off-by to this commit: f6777abdc4c002bfdbb314f56bcf06f0e327857d

Signed-off-by: sewon.jeon <irocks0922@gmail.com>

eclipse0922 commented:
@ericspod 
I agree with the concern about transform complexity increasing as these fixes accumulate. For this PR, I kept the scope limited to the regression fix and tests so we can unblock the issue safely.

As follow-up work, I will open a separate refactor thread to reduce complexity in this area (without changing behavior):

centralize inverse transform selection/filtering logic in one shared path for both MetaTensor and trace-key data,
reduce implicit grouping mechanics by moving toward a clearer, explicit tracking flow,
add focused regression coverage around multi-pipeline and mixed post-processing cases.
This should let us keep this bugfix small while improving maintainability in the next step.

eclipse0922 committed:
Refactor error messages for clarity and consistency in dictionary transforms

DCO Remediation Commit for sewon jeon <irocks0922@gmail.com>

I, sewon jeon <irocks0922@gmail.com>, hereby add my Signed-off-by to this commit: 4a918610715c9c15bb12315f6bdbcd1c148ad098
I, sewon jeon <irocks0922@gmail.com>, hereby add my Signed-off-by to this commit: f1f92e02403bbd9ae27300dfd5ca57aa9794e0a6

Signed-off-by: sewon jeon <irocks0922@gmail.com>

eclipse0922 committed:
DCO Remediation Commit for sewon.jeon <irocks0922@gmail.com>

I, sewon.jeon <irocks0922@gmail.com>, hereby add my Signed-off-by to this commit: 95222f3e89a0dbeb5d5d752f8702c2308a5dc96b

Signed-off-by: sewon.jeon <irocks0922@gmail.com>

eclipse0922 committed:
DCO Remediation Commit for sewon jeon <irocks0922@gmail.com>

I, sewon jeon <irocks0922@gmail.com>, hereby add my Signed-off-by to this commit: 8b31f24d3b509ae86fc2184a314004af24a79444

Signed-off-by: sewon jeon <irocks0922@gmail.com>

eclipse0922 committed:
DCO Remediation Commit for sewon.jeon <irocks0922@gmail.com>

I, sewon.jeon <irocks0922@gmail.com>, hereby add my Signed-off-by to this commit: 95222f3e89a0dbeb5d5d752f8702c2308a5dc96b

Signed-off-by: sewon.jeon <irocks0922@gmail.com>

Commit:
DCO Remediation Commit for sewon.jeon <sewon.jeon@connecteve.com>

I, sewon.jeon <sewon.jeon@connecteve.com>, hereby add my Signed-off-by to this commit: a607e700327f679f59062a124f48ba4b84e35b32

Signed-off-by: sewon.jeon <sewon.jeon@connecteve.com>

eclipse0922 committed:
Apply black 25.11.0 formatting fixes

Signed-off-by: sewon.jeon <irocks0922@gmail.com>

atbenmurray commented:
Taking a look

atbenmurray commented:
Okay, so I've started reviewing this finally, my apologies for the delay.

atbenmurray commented:
So, I've had a deeper look:

Going back to the motivating example in #8396, I'd observe the following:

```python
from monai.data import MetaTensor, create_test_image_2d
from monai.transforms import Compose, EnsureChannelFirstd, Invertd, Spacingd
from monai.transforms.utility.dictionary import Lambdad

img, _ = create_test_image_2d(60, 60, 2, 10, num_seg_classes=2)
img = MetaTensor(img, meta={"original_channel_dim": float("nan"), "pixdim": [1.0, 1.0]})
key = "image"
preprocessing = Compose([
        EnsureChannelFirstd(key),
        Spacingd(key, pixdim=[2.0, 2.0]),
])

postprocessing = Compose([
    Lambdad(key, func=lambda x: x),  # adds something to the applied operations of img
    Invertd(key, transform=preprocessing, orig_keys=key)  # gets confused by this item
])

item = {key: img}
pre = preprocessing(item)
post = postprocessing(pre)  # exception raised here
```

There seems to be some design tension between `Invertd` taking a sequence of transforms to be inverted relative to the logging of invertable transforms as applied operations. Ideally, we shouldn't have to pass the transforms at all, if all the required information is logged in the operations that have been performed on the data, but I'll proceed assuming we need them for now.

A cleaner approach might be to tweak the invert process so that it skips entries in the applied operations list that don't match the entries in the transforms passed to `Invertd`. I think that would achieve the same effect as the current PR without having to introduce a new group concept.

If we were to modify `pop_transform` and `get_most_recent_transform` so that each call to `pop_transform` pops the corresponding entry in `applied_operations` rather than requiring it to be the first entry in the list. It has the same effect as filtering on groups except it doesn't require groups


What Do You Think?

eclipse0922 commented:
@atbenmurray I checked this more closely and I agree with your direction here.

The current group-based change fixes the motivating `Invertd` path from `#8396`, but it is adding complexity around tracing/filtering instead of addressing the inverse-selection problem directly. It also leaves `Compose.inverse()` on mixed histories relying on the original stack ordering.

I'm going to rework this PR in the direction you suggested: drop the group plumbing and move the fix into the inverse stack handling so we match/pop the entries corresponding to the transform sequence being inverted, while still failing if an expected entry cannot be found.

That keeps the scope on `#8396`, but moves the fix closer to the root cause.

eclipse0922 committed:
Fix inverse matching with mixed postprocessing history

Signed-off-by: sewon.jeon <irocks0922@gmail.com>

eclipse0922 committed:
Add same-class inverse history regression test

Signed-off-by: sewon.jeon <irocks0922@gmail.com>

eclipse0922 committed:
Handle unset start method in inverse matching

Signed-off-by: sewon.jeon <irocks0922@gmail.com>

eclipse0922 committed:
Merge branch 'dev' into fix_invertd

atbenmurray reviewed (approved):
This seems much cleaner. It would be worth changing the docstring for `check_transforms_match` as the ID isn't used when running multiprocessing, and it might be worth adding a docstring to _transforms_match, but code wise it looks good

eclipse0922 committed:
Clarify inverse matching docs and apply formatting

Signed-off-by: sewon jeon <irocks0922@gmail.com>

eclipse0922 committed:
fix format

DCO Remediation Commit for sewon.jeon <irocks0922@gmail.com>

I, sewon.jeon <irocks0922@gmail.com>, hereby add my Signed-off-by to this commit: 165ff8a1e24625363898c445e3b259fbb5d7a8eb

Signed-off-by: sewon.jeon <irocks0922@gmail.com>

eclipse0922 commented:
Hi @ericspod, the PR has been reworked based on @atbenmurray's suggestion (dropped the group approach, 
moved the fix into inverse stack matching). @atbenmurray has approved the updated version. 
Could you take another look when you get a chance? Thanks

ericspod committed:
Merge branch 'dev' into fix_invertd

ericspod reviewed (approved):
I think we're good now if both me and @atbenmurray have had a look. Thanks!
